x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java

```diff
@@ -6,25 +6,153 @@
 
 package org.elasticsearch.repositories.encrypted;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.metadata.RepositoryMetadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.SecureSetting;
+import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.common.settings.Setting;
-import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.license.LicenseUtils;
+import org.elasticsearch.license.XPackLicenseState;
 import org.elasticsearch.plugins.Plugin;
-import org.elasticsearch.plugins.ReloadablePlugin;
 import org.elasticsearch.plugins.RepositoryPlugin;
+import org.elasticsearch.repositories.Repository;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.xpack.core.XPackPlugin;
 
+import java.security.GeneralSecurityException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
 
-public final class EncryptedRepositoryPlugin extends Plugin implements RepositoryPlugin, ReloadablePlugin {
+public class EncryptedRepositoryPlugin extends Plugin implements RepositoryPlugin {
+    static final Logger logger = LogManager.getLogger(EncryptedRepositoryPlugin.class);
+    static final String REPOSITORY_TYPE_NAME = "encrypted";
+    // TODO add at least hdfs, and investigate supporting all `BlobStoreRepository` implementations
+    static final List<String> SUPPORTED_ENCRYPTED_TYPE_NAMES = Arrays.asList("fs", "gcs", "azure", "s3");
+    static final Setting.AffixSetting<SecureString> ENCRYPTION_PASSWORD_SETTING = Setting.affixKeySetting(
+        "repository.encrypted.",
+        "password",
+        key -> SecureSetting.secureString(key, null)
+    );
+    static final Setting<String> DELEGATE_TYPE_SETTING = Setting.simpleString("delegate_type", "");
+    static final Setting<String> PASSWORD_NAME_SETTING = Setting.simpleString("password_name", "");
 
-    public EncryptedRepositoryPlugin(final Settings settings) {}
+    // "protected" because it is overloaded for tests
+    protected XPackLicenseState getLicenseState() {
+        return XPackPlugin.getSharedLicenseState();
+    }
+
+    public EncryptedRepositoryPlugin() {
+        if (false == getLicenseState().isEncryptedSnapshotAllowed()) {
+            logger.warn(
+                "Snapshotting to an encrypted repository is not permitted for the current license."
+                    + "All the other operations over the encrypted repository, eg. restore, work without restrictions.",
+                LicenseUtils.newComplianceException("encrypted snapshots")
+            );
```

Comment:
Yeah, you're right, this is silly. I suppose I wanted to warn users that encrypted repos are only partially supported, before actually creating them, but that's silly because you cannot tell if encrypted repos are gonna be used at all.